core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -323,8 +325,23 @@ private void updateStream() throws IOException {
     /* Create a new stream to read from mPosition. */
     // Calculate block id.
     long blockId = mStatus.getBlockIds().get(Math.toIntExact(mPosition / mBlockSize));
+    BlockInfo blockInfo = mStatus.getBlockInfo(blockId);
+    if (blockInfo == null) {
+      throw new IOException("No BlockInfo for block(id=" + blockId + ") of file"
+          + "(id=" + mStatus.getFileId() + ", path=" + mStatus.getPath() + ")");
+    }
     // Create stream
-    mBlockInStream = mBlockStore.getInStream(blockId, mOptions, mFailedWorkers);
+    boolean isBlockInfoOutdated = true;
+    for (BlockLocation location : blockInfo.getLocations()) {
```

Comment:
I'm confused by the logic. Shouldn't it be:
```java
boolean isBlockInfoOutdated = false;
for (BlockLocation location : blockInfo.getLocations()) {
if (mFailedWorkers.containsKey(location.getWorkerAddress())) {
        isBlockInfoOutdated = true;
        break;
      }
  }
```

Reply:
hrmmm, yeah, I'm not sure what this is doing either. Is there a test case which covers this code?

Reply:
blockInfo is "outdated" when all the locations in that blockInfo are failed workers, if there is at least one location that is not a failed worker, then it's not outdated.

Reply:
without this logic, `alluxio.server.ft.MultiWorkerIntegrationTest.readRecoverFromLostWorker` and `alluxio.server.ft.MultiWorkerIntegrationTest.readOneRecoverFromLostWorker` will fail.

Reply:
I see. Can you add that explanation as a comment? Also, once `isBlockInfoOutdated = false`, you can `break` out of the loop, right?